### ISSUE TITLE
Memory leaks transmission-daemon 3.00

### ISSUE THREAD

**Issue by bwesterb** [transmission/transmission#3055]
Just upgraded to 3.00 (bb6b5a062e) on Ubuntu and am suffering memory leaks. Have to restart the service every few hours to prevent it ballooning out of control. I checked process maps and its anonymous mmaps, so leaking heap allocations.

I see there are memory leaks fixed since 3.00: #2851 #2675 #2481 #2293 #1222

As others are experiencing issues as well, eg. #1379, I think it would be nice to have a new release.

<ins>(New release is discussed in #2287)</ins>

**Comment by bwesterb** [comment]
For those running into the same thing, the work around I use on Ubuntu is to create a new file

`/etc/systemd/system/transmission-daemon.service.d/restart.conf`

with contents

```
[Service]
Restart=always
RuntimeMaxSec=28800
MemoryMax=1G
```

which will restart transmission-daemon every 8 hours or when it reaches 1GB of memory. Don't forget to run `systemctl daemon-reload`.

**Comment by ckerr** [comment]
Agreed that the next step would be to have a new release.

That's not a procedural issue, not a bug to be tracked and fixes, so I'm going to close this as a duplicate of #2287.

If either you or the bug reporter in #3077 try out a nightly build or build from source and have more information on the behavior of the system there, please open a new issue if necessary.

**Comment by hasezoey** [comment]
continuing from #3077, i have tried to build, install and run transmission git, but i am unable to start the new version as a service (times out):

```txt
$ sudo service transmission-daemon.serivce status
May 13 23:35:31 sylvi systemd[1]: Starting Transmission BitTorrent Daemon...
May 13 23:35:31 sylvi transmission-daemon[369556]: [2022-05-13 23:35:31.452] utils.cc:212: Couldn't read '/var/lib/transmission-daemon/.config/user-dirs.dirs': No such file or directory (2)
May 13 23:37:01 sylvi systemd[1]: transmission-daemon.service: start operation timed out. Terminating.
May 13 23:37:22 sylvi transmission-daemon[369556]: Closing transmission session... done.
May 13 23:37:22 sylvi systemd[1]: transmission-daemon.service: Failed with result 'timeout'.
May 13 23:37:22 sylvi systemd[1]: Failed to start Transmission BitTorrent Daemon.
May 13 23:37:22 sylvi systemd[1]: transmission-daemon.service: Consumed 2.025s CPU time.
```

```txt
$ transmission-daemon --version
transmission-daemon --version
[2022-05-13 23:38:38.128] utils.cc:212: Couldn't read '/home/hasezoey/.config/user-dirs.dirs': No such file or directory (2)
[2022-05-13 23:38:38.128] utils.cc:212: Couldn't read '/home/hasezoey/.config/transmission-daemon/settings.json': No such file or directory (2)
transmission-daemon 3.00+ (5026d43920)
```

```txt
[Unit]
Description=Transmission BitTorrent Daemon
After=network.target

[Service]
User=debian-transmission
Type=notify
ExecStart=/usr/local/bin/transmission-daemon -f --log-level=error
ExecStop=/bin/kill -s STOP $MAINPID
ExecReload=/bin/kill -s HUP $MAINPID
NoNewPrivileges=true

[Install]
WantedBy=multi-user.target
```

result when manually running:
```txt
$ sudo /usr/local/bin/transmission-daemon -f --log-level=info 2>&1 | tee transmission_log
[2022-05-13 23:46:51.785] utils.cc:212: Couldn't read '/root/.config/user-dirs.dirs': No such file or directory (2)
[2022-05-13 23:46:52.889] inf session.cc:753 Transmission version 3.00+ (5026d43920) starting (/custom/transmission/libtransmission/session.cc:753)
[2022-05-13 23:46:52.889] ERR utils.cc:212 Couldn't read '/root/.config/transmission-daemon/stats.json': No such file or directory (2) (/custom/transmission/libtransmission/utils.cc:212)
[2022-05-13 23:46:52.889] ERR utils.cc:212 Couldn't read '/root/.config/transmission-daemon/stats.benc': No such file or directory (2) (/custom/transmission/libtransmission/utils.cc:212)
[2022-05-13 23:46:52.889] inf rpc-server.cc:875 Added '127.0.0.1' to host whitelist (/custom/transmission/libtransmission/rpc-server.cc:875)
[2022-05-13 23:46:52.889] inf rpc-server.cc:875 Added '::1' to host whitelist (/custom/transmission/libtransmission/rpc-server.cc:875)
[2022-05-13 23:46:52.889] inf rpc-server.cc:1134 Serving RPC and Web requests on 0.0.0.0:9091/transmission/ (/custom/transmission/libtransmission/rpc-server.cc:1134)
[2022-05-13 23:46:52.889] inf rpc-server.cc:773 Listening for RPC and Web requests on '0.0.0.0:9091' (/custom/transmission/libtransmission/rpc-server.cc:773)
[2022-05-13 23:46:52.889] inf rpc-server.cc:1139 Whitelist enabled (/custom/transmission/libtransmission/rpc-server.cc:1139)
[2022-05-13 23:46:52.889] inf rpc-server.cc:1151 Serving RPC and Web requests from '/usr/local/share/transmission/public_html' (/custom/transmission/libtransmission/rpc-server.cc:1151)
[2022-05-13 23:46:52.889] inf tr-dht.cc:287 Initializing DHT (/custom/transmission/libtransmission/tr-dht.cc:287)
[2022-05-13 23:46:52.889] ERR utils.cc:212 Couldn't read '/root/.config/transmission-daemon/dht.dat': No such file or directory (2) (/custom/transmission/libtransmission/utils.cc:212)
[2022-05-13 23:46:52.889] inf web.cc:124 Will verify tracker certs using envvar CURL_CA_BUNDLE: none (/custom/transmission/libtransmission/web.cc:124)
[2022-05-13 23:46:52.889] inf web.cc:126 NB: this only works if you built against libcurl with openssl or gnutls, NOT nss (/custom/transmission/libtransmission/web.cc:126)
[2022-05-13 23:46:52.889] inf web.cc:127 NB: Invalid certs will appear as 'Could not connect to tracker' like many other errors (/custom/transmission/libtransmission/web.cc:127)
[2022-05-13 23:46:52.889] inf daemon.cc:746 Loading settings from '/root/.config/transmission-daemon' (/custom/transmission/daemon/daemon.cc:746)
[2022-05-13 23:46:52.889] inf port-forwarding.cc:110 State changed from 'Not forwarded' to 'Starting' (/custom/transmission/libtransmission/port-forwarding.cc:110)

(after ~5min i used CTRL+C because nothing was happening)
```

PS: i had build with `libcurl4-openssl-dev` because i didnt know what to choose from the 3 options

**Comment by hasezoey** [comment]
> @hasezoey just wondering, did you download the source zip from Github?

no, i use the git repo, with all the instructions from https://github.com/transmission/transmission#building-transmission-from-git-updating (updating since i had it since my first build comment)

**Comment by hasezoey** [comment]
> I'd try the 'git submodule update --recursive' again. I can see i got fast_float in the third-party directory correctly.

just tried again with all the steps, but same error

**Comment by GaryElshaw** [comment]
is 'fast_float.h' in Transmission/third-party/fast_float/include/fast_float ?

**Comment by hasezoey** [comment]
> is 'fast_float.h' in Transmission/third-party/fast_float/include/fast_float ?

if you mean with `Transmission` repository root, then the `third-party/fast_float` directory exists, but is empty

**Comment by hasezoey** [comment]
> Sometimes a fresh clone is the best way to go.

can confirm that newly cloning somehow fixed it, is it just that git submodules are weird?

**Comment by hasezoey** [comment]
after having resolved the build issue, testing with e286e45ac848dc0c3497390fc81a47b51890902b again, it seems to run this time but systemd still has a timeout trying to start the service (though i can connect to it via transmission-remote-gtk)

service file:
```txt
[Unit]
Description=Transmission BitTorrent Daemon
After=network.target

[Service]
User=debian-transmission
Type=notify
ExecStart=/usr/local/bin/transmission-daemon -f --log-level=info
ExecStop=/bin/kill -s STOP $MAINPID
ExecReload=/bin/kill -s HUP $MAINPID
NoNewPrivileges=true

[Install]
WantedBy=multi-user.target
```

service log:
```txt
May 16 14:54:13 sylvi systemd[1]: transmission-daemon.service: start operation timed out. Terminating.
May 16 14:54:34 sylvi transmission-daemon[625354]: [2022-05-16 14:54:34.014] inf session.cc:1951 Transmission version 3.00+ (e286e45ac8) shutting down (/cu>
May 16 14:54:34 sylvi transmission-daemon[625354]: [2022-05-16 14:54:34.014] inf rpc-server.cc:801 Stopped listening for RPC and Web requests on '0.0.0.0:9>
May 16 14:54:34 sylvi transmission-daemon[625354]: Closing transmission session... done.
May 16 14:54:34 sylvi systemd[1]: transmission-daemon.service: Failed with result 'timeout'.
May 16 14:54:34 sylvi systemd[1]: Failed to start Transmission BitTorrent Daemon.
May 16 14:54:34 sylvi systemd[1]: transmission-daemon.service: Consumed 1.869s CPU time.
```

**Comment by hasezoey** [comment]
as a small update, i got it to run with systemd, i was not familiar enough with systemd or cmake, but have found that i need to add the cmake option `-DWITH_SYSTEMD=ON` and have `libsystemd-dev` installed, or have changed `Type=notify` to `Type=simple` in the service file

as for now (running ~10min) it had not increased in memory yet beyond what i would expect
